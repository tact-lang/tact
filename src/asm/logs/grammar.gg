vmLine =
      VmLoc
    / VmStack
    / VmExecute
    / VmLimitChanged
    / VmGasRemaining
    / VmException
    / VmExceptionHandler
    / VmFinalC5
    / VmUnknown;

VmStack            = "stack: " stack:$[^\n]+;
VmLoc              = "code cell hash:" hash:hex "offset:" offset:number;
VmExecute          = "execute " instr:$[^\n]+;
VmLimitChanged     = "changing gas limit to " limit:number;
VmGasRemaining     = "gas remaining: " gas:number;
VmException        = "handling exception code " errno:number ": " message:$[^\n]*;
VmExceptionHandler = "default exception handler, terminating vm with exit code" errno:number;
VmFinalC5          = "final c5:" value:Cell;
VmUnknown          = !"stack" text:$[^\n]+ "\n";

VmParsedStack = "[" values:VmStackValue* "]";

VmStackValue = value:(
      Null
    / NaN
    / Integer
    / Tuple
    / TupleParen
    / Cell
    / Continuation
    / Builder
    / CellSlice
    / Unknown
);

Null         = "()" / "(null)";
NaN          = "NaN";
Integer      = value:number;
Tuple        = "[" elements:VmStackValue* "]";
TupleParen   = "(" elements:VmStackValue* ")";
Cell         = "C{" value:hex "}";
Continuation = "Cont{" value:$#([A-Za-z_0-9]*) "}";
Builder      = "BC{" value:hex "}";
Unknown      = "???";

CellSlice     = "CS{" body:(CellSliceBody / CellSliceShortBody) "}";
CellSliceBody = "Cell{" value:hex "}" bits:CellSliceBits ";" refs:CellSliceRefs;
CellSliceBits = "bits:" start:number ".." end:number;
CellSliceRefs = "refs:" start:number ".." end:number;

CellSliceShortBody = value:hex;

number   = op:"-"? value:#$(digit+);
digit    = [0-9];

hexDigit "hexadecimal digit" = [0-9a-fA-F];
hex      = $#(value:hexDigit*);

space "space" = [ \t\r\n];
