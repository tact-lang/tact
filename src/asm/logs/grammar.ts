/* Generated. Do not edit. */
/* eslint-disable @typescript-eslint/no-namespace */
/* eslint-disable @typescript-eslint/ban-types */
/* eslint-disable @typescript-eslint/no-redundant-type-constituents */
/* eslint-disable @typescript-eslint/no-duplicate-type-constituents */
/* eslint-disable @typescript-eslint/no-unused-vars */
import * as $ from "@tonstudio/parser-runtime"
export namespace $ast {
    export type VmLoc = $.Located<{
        readonly $: "VmLoc"
        readonly hash: hex
        readonly offset: $number
    }>
    export type VmStack = $.Located<{
        readonly $: "VmStack"
        readonly stack: string
    }>
    export type VmExecute = $.Located<{
        readonly $: "VmExecute"
        readonly instr: string
    }>
    export type VmLimitChanged = $.Located<{
        readonly $: "VmLimitChanged"
        readonly limit: $number
    }>
    export type VmGasRemaining = $.Located<{
        readonly $: "VmGasRemaining"
        readonly gas: $number
    }>
    export type VmException = $.Located<{
        readonly $: "VmException"
        readonly errno: $number
        readonly message: string
    }>
    export type VmExceptionHandler = $.Located<{
        readonly $: "VmExceptionHandler"
        readonly errno: $number
    }>
    export type VmFinalC5 = $.Located<{
        readonly $: "VmFinalC5"
        readonly value: Cell
    }>
    export type VmUnknown = $.Located<{
        readonly $: "VmUnknown"
        readonly text: string
    }>
    export type vmLine =
        | VmLoc
        | VmStack
        | VmExecute
        | VmLimitChanged
        | VmGasRemaining
        | VmException
        | VmExceptionHandler
        | VmFinalC5
        | VmUnknown
    export type VmParsedStack = $.Located<{
        readonly $: "VmParsedStack"
        readonly values: readonly VmStackValue[]
    }>
    export type VmStackValue = $.Located<{
        readonly $: "VmStackValue"
        readonly value:
            | Null
            | NaN
            | Integer
            | Tuple
            | TupleParen
            | Cell
            | Continuation
            | Builder
            | CellSlice
            | Unknown
    }>
    export type Null = $.Located<{
        readonly $: "Null"
    }>
    export type NaN = $.Located<{
        readonly $: "NaN"
    }>
    export type Integer = $.Located<{
        readonly $: "Integer"
        readonly value: $number
    }>
    export type Tuple = $.Located<{
        readonly $: "Tuple"
        readonly elements: readonly VmStackValue[]
    }>
    export type TupleParen = $.Located<{
        readonly $: "TupleParen"
        readonly elements: readonly VmStackValue[]
    }>
    export type Cell = $.Located<{
        readonly $: "Cell"
        readonly value: hex
    }>
    export type Continuation = $.Located<{
        readonly $: "Continuation"
        readonly value: string
    }>
    export type Builder = $.Located<{
        readonly $: "Builder"
        readonly value: hex
    }>
    export type Unknown = $.Located<{
        readonly $: "Unknown"
    }>
    export type CellSlice = $.Located<{
        readonly $: "CellSlice"
        readonly body: CellSliceBody | CellSliceShortBody
    }>
    export type CellSliceBody = $.Located<{
        readonly $: "CellSliceBody"
        readonly value: hex
        readonly bits: CellSliceBits
        readonly refs: CellSliceRefs
    }>
    export type CellSliceBits = $.Located<{
        readonly $: "CellSliceBits"
        readonly start: $number
        readonly end: $number
    }>
    export type CellSliceRefs = $.Located<{
        readonly $: "CellSliceRefs"
        readonly start: $number
        readonly end: $number
    }>
    export type CellSliceShortBody = $.Located<{
        readonly $: "CellSliceShortBody"
        readonly value: hex
    }>
    export type $number = {
        readonly op: "-" | undefined
        readonly value: string
    }
    export type digit = string
    export type hexDigit = string | string | string
    export type hex = string
    export type space = " " | "\t" | "\r" | "\n"
}
export const VmLoc: $.Parser<$ast.VmLoc> = $.loc(
    $.field(
        $.pure("VmLoc"),
        "$",
        $.right(
            $.str("code cell hash:"),
            $.field(
                $.lazy(() => hex),
                "hash",
                $.right(
                    $.str("offset:"),
                    $.field(
                        $.lazy(() => $number),
                        "offset",
                        $.eps,
                    ),
                ),
            ),
        ),
    ),
)
export const VmStack: $.Parser<$ast.VmStack> = $.loc(
    $.field(
        $.pure("VmStack"),
        "$",
        $.right(
            $.str("stack: "),
            $.field(
                $.stry($.plus($.regex<"\n">("^\\n", $.negateExps([$.ExpString("\n")])))),
                "stack",
                $.eps,
            ),
        ),
    ),
)
export const VmExecute: $.Parser<$ast.VmExecute> = $.loc(
    $.field(
        $.pure("VmExecute"),
        "$",
        $.right(
            $.str("execute "),
            $.field(
                $.stry($.plus($.regex<"\n">("^\\n", $.negateExps([$.ExpString("\n")])))),
                "instr",
                $.eps,
            ),
        ),
    ),
)
export const VmLimitChanged: $.Parser<$ast.VmLimitChanged> = $.loc(
    $.field(
        $.pure("VmLimitChanged"),
        "$",
        $.right(
            $.str("changing gas limit to "),
            $.field(
                $.lazy(() => $number),
                "limit",
                $.eps,
            ),
        ),
    ),
)
export const VmGasRemaining: $.Parser<$ast.VmGasRemaining> = $.loc(
    $.field(
        $.pure("VmGasRemaining"),
        "$",
        $.right(
            $.str("gas remaining: "),
            $.field(
                $.lazy(() => $number),
                "gas",
                $.eps,
            ),
        ),
    ),
)
export const VmException: $.Parser<$ast.VmException> = $.loc(
    $.field(
        $.pure("VmException"),
        "$",
        $.right(
            $.str("handling exception code "),
            $.field(
                $.lazy(() => $number),
                "errno",
                $.right(
                    $.str(": "),
                    $.field(
                        $.stry($.star($.regex<"\n">("^\\n", $.negateExps([$.ExpString("\n")])))),
                        "message",
                        $.eps,
                    ),
                ),
            ),
        ),
    ),
)
export const VmExceptionHandler: $.Parser<$ast.VmExceptionHandler> = $.loc(
    $.field(
        $.pure("VmExceptionHandler"),
        "$",
        $.right(
            $.str("default exception handler, terminating vm with exit code"),
            $.field(
                $.lazy(() => $number),
                "errno",
                $.eps,
            ),
        ),
    ),
)
export const VmFinalC5: $.Parser<$ast.VmFinalC5> = $.loc(
    $.field(
        $.pure("VmFinalC5"),
        "$",
        $.right(
            $.str("final c5:"),
            $.field(
                $.lazy(() => Cell),
                "value",
                $.eps,
            ),
        ),
    ),
)
export const VmUnknown: $.Parser<$ast.VmUnknown> = $.loc(
    $.field(
        $.pure("VmUnknown"),
        "$",
        $.right(
            $.lookNeg($.str("stack")),
            $.field(
                $.stry($.plus($.regex<"\n">("^\\n", $.negateExps([$.ExpString("\n")])))),
                "text",
                $.right($.str("\n"), $.eps),
            ),
        ),
    ),
)
export const vmLine: $.Parser<$ast.vmLine> = $.alt(
    VmLoc,
    $.alt(
        VmStack,
        $.alt(
            VmExecute,
            $.alt(
                VmLimitChanged,
                $.alt(
                    VmGasRemaining,
                    $.alt(VmException, $.alt(VmExceptionHandler, $.alt(VmFinalC5, VmUnknown))),
                ),
            ),
        ),
    ),
)
export const VmParsedStack: $.Parser<$ast.VmParsedStack> = $.loc(
    $.field(
        $.pure("VmParsedStack"),
        "$",
        $.right(
            $.str("["),
            $.field($.star($.lazy(() => VmStackValue)), "values", $.right($.str("]"), $.eps)),
        ),
    ),
)
export const VmStackValue: $.Parser<$ast.VmStackValue> = $.loc(
    $.field(
        $.pure("VmStackValue"),
        "$",
        $.field(
            $.alt(
                $.lazy(() => Null),
                $.alt(
                    $.lazy(() => NaN),
                    $.alt(
                        $.lazy(() => Integer),
                        $.alt(
                            $.lazy(() => Tuple),
                            $.alt(
                                $.lazy(() => TupleParen),
                                $.alt(
                                    $.lazy(() => Cell),
                                    $.alt(
                                        $.lazy(() => Continuation),
                                        $.alt(
                                            $.lazy(() => Builder),
                                            $.alt(
                                                $.lazy(() => CellSlice),
                                                $.lazy(() => Unknown),
                                            ),
                                        ),
                                    ),
                                ),
                            ),
                        ),
                    ),
                ),
            ),
            "value",
            $.eps,
        ),
    ),
)
export const Null: $.Parser<$ast.Null> = $.loc(
    $.field($.pure("Null"), "$", $.right($.alt($.str("()"), $.str("(null)")), $.eps)),
)
export const NaN: $.Parser<$ast.NaN> = $.loc(
    $.field($.pure("NaN"), "$", $.right($.str("NaN"), $.eps)),
)
export const Integer: $.Parser<$ast.Integer> = $.loc(
    $.field(
        $.pure("Integer"),
        "$",
        $.field(
            $.lazy(() => $number),
            "value",
            $.eps,
        ),
    ),
)
export const Tuple: $.Parser<$ast.Tuple> = $.loc(
    $.field(
        $.pure("Tuple"),
        "$",
        $.right($.str("["), $.field($.star(VmStackValue), "elements", $.right($.str("]"), $.eps))),
    ),
)
export const TupleParen: $.Parser<$ast.TupleParen> = $.loc(
    $.field(
        $.pure("TupleParen"),
        "$",
        $.right($.str("("), $.field($.star(VmStackValue), "elements", $.right($.str(")"), $.eps))),
    ),
)
export const Cell: $.Parser<$ast.Cell> = $.loc(
    $.field(
        $.pure("Cell"),
        "$",
        $.right(
            $.str("C{"),
            $.field(
                $.lazy(() => hex),
                "value",
                $.right($.str("}"), $.eps),
            ),
        ),
    ),
)
export const Continuation: $.Parser<$ast.Continuation> = $.loc(
    $.field(
        $.pure("Continuation"),
        "$",
        $.right(
            $.str("Cont{"),
            $.field(
                $.stry(
                    $.lex(
                        $.star(
                            $.regex<string | string | "_" | string>("A-Za-z_0-9", [
                                $.ExpRange("A", "Z"),
                                $.ExpRange("a", "z"),
                                $.ExpString("_"),
                                $.ExpRange("0", "9"),
                            ]),
                        ),
                    ),
                ),
                "value",
                $.right($.str("}"), $.eps),
            ),
        ),
    ),
)
export const Builder: $.Parser<$ast.Builder> = $.loc(
    $.field(
        $.pure("Builder"),
        "$",
        $.right(
            $.str("BC{"),
            $.field(
                $.lazy(() => hex),
                "value",
                $.right($.str("}"), $.eps),
            ),
        ),
    ),
)
export const Unknown: $.Parser<$ast.Unknown> = $.loc(
    $.field($.pure("Unknown"), "$", $.right($.str("???"), $.eps)),
)
export const CellSlice: $.Parser<$ast.CellSlice> = $.loc(
    $.field(
        $.pure("CellSlice"),
        "$",
        $.right(
            $.str("CS{"),
            $.field(
                $.alt(
                    $.lazy(() => CellSliceBody),
                    $.lazy(() => CellSliceShortBody),
                ),
                "body",
                $.right($.str("}"), $.eps),
            ),
        ),
    ),
)
export const CellSliceBody: $.Parser<$ast.CellSliceBody> = $.loc(
    $.field(
        $.pure("CellSliceBody"),
        "$",
        $.right(
            $.str("Cell{"),
            $.field(
                $.lazy(() => hex),
                "value",
                $.right(
                    $.str("}"),
                    $.field(
                        $.lazy(() => CellSliceBits),
                        "bits",
                        $.right(
                            $.str(";"),
                            $.field(
                                $.lazy(() => CellSliceRefs),
                                "refs",
                                $.eps,
                            ),
                        ),
                    ),
                ),
            ),
        ),
    ),
)
export const CellSliceBits: $.Parser<$ast.CellSliceBits> = $.loc(
    $.field(
        $.pure("CellSliceBits"),
        "$",
        $.right(
            $.str("bits:"),
            $.field(
                $.lazy(() => $number),
                "start",
                $.right(
                    $.str(".."),
                    $.field(
                        $.lazy(() => $number),
                        "end",
                        $.eps,
                    ),
                ),
            ),
        ),
    ),
)
export const CellSliceRefs: $.Parser<$ast.CellSliceRefs> = $.loc(
    $.field(
        $.pure("CellSliceRefs"),
        "$",
        $.right(
            $.str("refs:"),
            $.field(
                $.lazy(() => $number),
                "start",
                $.right(
                    $.str(".."),
                    $.field(
                        $.lazy(() => $number),
                        "end",
                        $.eps,
                    ),
                ),
            ),
        ),
    ),
)
export const CellSliceShortBody: $.Parser<$ast.CellSliceShortBody> = $.loc(
    $.field(
        $.pure("CellSliceShortBody"),
        "$",
        $.field(
            $.lazy(() => hex),
            "value",
            $.eps,
        ),
    ),
)
export const $number: $.Parser<$ast.$number> = $.field(
    $.opt($.str("-")),
    "op",
    $.field($.lex($.stry($.plus($.lazy(() => digit)))), "value", $.eps),
)
export const digit: $.Parser<$ast.digit> = $.regex<string>("0-9", [$.ExpRange("0", "9")])
export const hexDigit: $.Parser<$ast.hexDigit> = $.named(
    "hexadecimal digit",
    $.regex<string | string | string>("0-9a-fA-F", [
        $.ExpRange("0", "9"),
        $.ExpRange("a", "f"),
        $.ExpRange("A", "F"),
    ]),
)
export const hex: $.Parser<$ast.hex> = $.stry($.lex($.field($.star(hexDigit), "value", $.eps)))
export const space: $.Parser<$ast.space> = $.named(
    "space",
    $.regex<" " | "\t" | "\r" | "\n">(" \\t\\r\\n", [
        $.ExpString(" "),
        $.ExpString("\t"),
        $.ExpString("\r"),
        $.ExpString("\n"),
    ]),
)
