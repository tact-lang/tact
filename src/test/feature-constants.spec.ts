import { toNano } from "@ton/core";
import { ContractSystem } from "@tact-lang/emulator";
import { __DANGER_resetNodeId } from "../grammar/ast";
import { ConstantTester } from "./features/output/constants_ConstantTester";

describe("feature-constants", () => {
    beforeEach(() => {
        __DANGER_resetNodeId();
    });
    it("should implement constants correctly", async () => {
        // Init
        const system = await ContractSystem.create();
        const treasure = system.treasure("treasure");
        const contract = system.open(await ConstantTester.fromInit());
        await contract.send(treasure, { value: toNano("10") }, null);
        await system.run();
        expect(contract).toMatchSnapshot();

        // Check methods
        expect(await contract.getSomething1()).toEqual(11n);
        expect(await contract.getSomething2()).toBeNull();
        expect(await contract.getSomething3()).toEqual(toNano("123"));
        expect(await contract.getSomething4()).toEqual(toNano("456"));
        expect(await contract.getSomething5()).toEqual("Hello world!");
        expect(await contract.getSomething6()).toEqual(10n);
        expect(await contract.getSomething7()).toEqual(5n);
        expect(await contract.getSomething8()).toEqual(4n);
        expect((await contract.getSomething9()).toRawString()).toEqual(
            "0:4a81708d2cf7b15a1b362fbf64880451d698461f52f05f145b36c08517d76873",
        );
        expect((await contract.getSomething10()).toRawString()).toEqual(
            "0:4a81708d2cf7b15a1b362fbf64880451d698461f52f05f145b36c08517d76873",
        );
        expect(await contract.getSomething11()).toEqual(88n);
        expect(await contract.getSomething12()).toEqual(-90n);
        expect(await contract.getSomething13()).toEqual(88n);
        expect(await contract.getSomething14()).toEqual(243n);
        expect(await contract.getSomething15()).toEqual(32n);
        expect(await contract.getGlobalConst()).toEqual(100n);
    });
});
