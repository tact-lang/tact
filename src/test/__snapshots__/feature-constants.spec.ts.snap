// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`feature-constants should implement constants correctly 1`] = `
ConstantTester {
  "abi": {
    "errors": {
      "10": {
        "message": "Dictionary error",
      },
      "128": {
        "message": "Null reference exception",
      },
      "129": {
        "message": "Invalid serialization prefix",
      },
      "13": {
        "message": "Out of gas error",
      },
      "130": {
        "message": "Invalid incoming message",
      },
      "131": {
        "message": "Constraints error",
      },
      "132": {
        "message": "Access denied",
      },
      "133": {
        "message": "Contract stopped",
      },
      "134": {
        "message": "Invalid argument",
      },
      "135": {
        "message": "Code of a contract was not found",
      },
      "136": {
        "message": "Invalid address",
      },
      "137": {
        "message": "Masterchain support is not enabled for this contract",
      },
      "2": {
        "message": "Stack undeflow",
      },
      "3": {
        "message": "Stack overflow",
      },
      "32": {
        "message": "Method ID not found",
      },
      "34": {
        "message": "Action is invalid or not supported",
      },
      "37": {
        "message": "Not enough TON",
      },
      "38": {
        "message": "Not enough extra-currencies",
      },
      "4": {
        "message": "Integer overflow",
      },
      "5": {
        "message": "Integer out of expected range",
      },
      "6": {
        "message": "Invalid opcode",
      },
      "7": {
        "message": "Type check error",
      },
      "8": {
        "message": "Cell overflow",
      },
      "9": {
        "message": "Cell underflow",
      },
    },
    "getters": [
      {
        "arguments": [],
        "name": "something1",
        "returnType": {
          "format": 257,
          "kind": "simple",
          "optional": false,
          "type": "int",
        },
      },
      {
        "arguments": [],
        "name": "something2",
        "returnType": {
          "format": 257,
          "kind": "simple",
          "optional": true,
          "type": "int",
        },
      },
      {
        "arguments": [],
        "name": "something3",
        "returnType": {
          "format": 257,
          "kind": "simple",
          "optional": false,
          "type": "int",
        },
      },
      {
        "arguments": [],
        "name": "something4",
        "returnType": {
          "format": 257,
          "kind": "simple",
          "optional": false,
          "type": "int",
        },
      },
      {
        "arguments": [],
        "name": "something5",
        "returnType": {
          "kind": "simple",
          "optional": false,
          "type": "string",
        },
      },
      {
        "arguments": [],
        "name": "something6",
        "returnType": {
          "format": 257,
          "kind": "simple",
          "optional": false,
          "type": "int",
        },
      },
      {
        "arguments": [],
        "name": "something7",
        "returnType": {
          "format": 257,
          "kind": "simple",
          "optional": false,
          "type": "int",
        },
      },
      {
        "arguments": [],
        "name": "something8",
        "returnType": {
          "format": 257,
          "kind": "simple",
          "optional": false,
          "type": "int",
        },
      },
      {
        "arguments": [],
        "name": "something9",
        "returnType": {
          "kind": "simple",
          "optional": false,
          "type": "address",
        },
      },
      {
        "arguments": [],
        "name": "something10",
        "returnType": {
          "kind": "simple",
          "optional": false,
          "type": "address",
        },
      },
      {
        "arguments": [],
        "name": "something11",
        "returnType": {
          "format": 257,
          "kind": "simple",
          "optional": false,
          "type": "int",
        },
      },
      {
        "arguments": [],
        "name": "something12",
        "returnType": {
          "format": 257,
          "kind": "simple",
          "optional": false,
          "type": "int",
        },
      },
      {
        "arguments": [],
        "name": "something13",
        "returnType": {
          "format": 257,
          "kind": "simple",
          "optional": false,
          "type": "int",
        },
      },
      {
        "arguments": [],
        "name": "something14",
        "returnType": {
          "format": 257,
          "kind": "simple",
          "optional": false,
          "type": "int",
        },
      },
      {
        "arguments": [],
        "name": "something15",
        "returnType": {
          "format": 257,
          "kind": "simple",
          "optional": false,
          "type": "int",
        },
      },
      {
        "arguments": [],
        "name": "globalConst",
        "returnType": {
          "format": 257,
          "kind": "simple",
          "optional": false,
          "type": "int",
        },
      },
    ],
    "receivers": [
      {
        "message": {
          "kind": "empty",
        },
        "receiver": "internal",
      },
    ],
    "types": [
      {
        "fields": [
          {
            "name": "code",
            "type": {
              "kind": "simple",
              "optional": false,
              "type": "cell",
            },
          },
          {
            "name": "data",
            "type": {
              "kind": "simple",
              "optional": false,
              "type": "cell",
            },
          },
        ],
        "header": null,
        "name": "StateInit",
      },
      {
        "fields": [
          {
            "name": "bounced",
            "type": {
              "kind": "simple",
              "optional": false,
              "type": "bool",
            },
          },
          {
            "name": "sender",
            "type": {
              "kind": "simple",
              "optional": false,
              "type": "address",
            },
          },
          {
            "name": "value",
            "type": {
              "format": 257,
              "kind": "simple",
              "optional": false,
              "type": "int",
            },
          },
          {
            "name": "raw",
            "type": {
              "kind": "simple",
              "optional": false,
              "type": "slice",
            },
          },
        ],
        "header": null,
        "name": "Context",
      },
      {
        "fields": [
          {
            "name": "bounce",
            "type": {
              "kind": "simple",
              "optional": false,
              "type": "bool",
            },
          },
          {
            "name": "to",
            "type": {
              "kind": "simple",
              "optional": false,
              "type": "address",
            },
          },
          {
            "name": "value",
            "type": {
              "format": 257,
              "kind": "simple",
              "optional": false,
              "type": "int",
            },
          },
          {
            "name": "mode",
            "type": {
              "format": 257,
              "kind": "simple",
              "optional": false,
              "type": "int",
            },
          },
          {
            "name": "body",
            "type": {
              "kind": "simple",
              "optional": true,
              "type": "cell",
            },
          },
          {
            "name": "code",
            "type": {
              "kind": "simple",
              "optional": true,
              "type": "cell",
            },
          },
          {
            "name": "data",
            "type": {
              "kind": "simple",
              "optional": true,
              "type": "cell",
            },
          },
        ],
        "header": null,
        "name": "SendParameters",
      },
    ],
  },
  "address": kQCG5ZO3hYCRKTIcSHbxVD6aemP9eqeRqXiPLuTtmDjr2Wio,
  "init": {
    "code": x{FF00F4A413F4BCF2C80B}
 x{62_}
  x{D001D0D3030171B0A301FA400120D74981010BBAF2E08820D70B0A208104FFBAF2D0898309BAF2E088545053036F04F86102F862DB3C59DB3CF2E08230C8F84301CC7F01CA00C9ED54}
   x{ED44D0D401F863D20030916DE0F828D70B0A8309BAF2E089DB3C}
    x{6D}
   x{0192307FE07021D749C21F953020D70B1FDEC00001D749C121B0917FE070}
  x{2_}
   x{6A_}
    x{B22EF6CF36CF0C6_}
     x{ED44D0D401F863D20030916DE0F828D70B0A8309BAF2E089DB3C}
      x{6D}
     x{74}
    x{B226B6CF36CF0C6_}
     x{ED44D0D401F863D20030916DE0F828D70B0A8309BAF2E089DB3C}
      x{6D}
     x{8D086002540B846967BD8AD0D9B17DFB2440228EB4C230FA9782F8A2D9B60428BEBB439C}
   x{2_}
    x{2_}
     x{2_}
      x{B24DF6CF36CF0C6_}
       x{ED44D0D401F863D20030916DE0F828D70B0A8309BAF2E089DB3C}
        x{6D}
       x{82186A2BB7D000}
      x{B245B6CF36CF0C6_}
       x{ED44D0D401F863D20030916DE0F828D70B0A8309BAF2E089DB3C}
        x{6D}
       x{8BC48656C6C6F20776F726C64218}
     x{2_}
      x{5}
       x{A975DB3CDB3C31}
        x{ED44D0D401F863D20030916DE0F828D70B0A8309BAF2E089DB3C}
         x{6D}
        x{7A}
       x{AAA2DB3CDB3C31}
        x{ED44D0D401F863D20030916DE0F828D70B0A8309BAF2E089DB3C}
         x{6D}
        x{8020}
      x{5}
       x{2_}
        x{A6A9B679B67863_}
         x{ED44D0D401F863D20030916DE0F828D70B0A8309BAF2E089DB3C}
          x{6D}
         x{75}
        x{A7A304E0B9D87ABA595CF63D09D873AC950A36CD04E13BAC6A5A56C8F96645A75E5BFBE2C47304E040AB803700CEC7299CB0755C8E4B658704E037A93BDF2A8B2DE38EA7A7BC0AB3864D04E13B2E9CB5669D96E6741D2CDB28CDD149_}
       x{AA83DB3CDB3C31}
        x{ED44D0D401F863D20030916DE0F828D70B0A8309BAF2E089DB3C}
         x{6D}
        x{8100F3}
    x{2_}
     x{2_}
      x{2_}
       x{2_}
        x{AABEED44D0D20001}
        x{AB61DB3CDB3C31}
         x{ED44D0D401F863D20030916DE0F828D70B0A8309BAF2E089DB3C}
          x{6D}
         x{8064}
       x{AF326D9E6D9E18C_}
        x{ED44D0D401F863D20030916DE0F828D70B0A8309BAF2E089DB3C}
         x{6D}
        x{8058}
      x{5}
       x{79A_}
        x{ACB6CF36CF0C6_}
         x{ED44D0D401F863D20030916DE0F828D70B0A8309BAF2E089DB3C}
          x{6D}
         x{800B}
        x{AEE3435697066733A2F2F516D625A38617576474E63656B454543537056507163476A37506868564C69697A724E384D364C515248526E6A5182_}
       x{AA45DB3CDB3C31}
        x{ED44D0D401F863D20030916DE0F828D70B0A8309BAF2E089DB3C}
         x{6D}
        x{80A6}
     x{2_}
      x{5}
       x{A9F1DB3CDB3C31}
        x{ED44D0D401F863D20030916DE0F828D70B0A8309BAF2E089DB3C}
         x{6D}
        x{6D}
       x{AA26DB3CDB3C31}
        x{ED44D0D401F863D20030916DE0F828D70B0A8309BAF2E089DB3C}
         x{6D}
        x{8058}
      x{5}
       x{A9D0DB3CDB3C31}
        x{ED44D0D401F863D20030916DE0F828D70B0A8309BAF2E089DB3C}
         x{6D}
        x{82181CA35F0E00}
       x{AA07DB3CDB3C31}
        x{ED44D0D401F863D20030916DE0F828D70B0A8309BAF2E089DB3C}
         x{6D}
        x{8D086002540B846967BD8AD0D9B17DFB2440228EB4C230FA9782F8A2D9B60428BEBB439C},
    "data": x{4_}
 x{C_}
  x{A0F903_}
   x{FF00F4A413F4BCF2C80B}
    x{62_}
     x{D001D0D3030171B0A301FA400120D74981010BBAF2E08820D70B0A208104FFBAF2D0898309BAF2E088545053036F04F86102F862DB3C59DB3CF2E08230C8F84301CC7F01CA00C9ED54}
      x{ED44D0D401F863D20030916DE0F828D70B0A8309BAF2E089DB3C}
       x{6D}
      x{0192307FE07021D749C21F953020D70B1FDEC00001D749C121B0917FE070}
     x{2_}
      x{6A_}
       x{B22EF6CF36CF0C6_}
        x{ED44D0D401F863D20030916DE0F828D70B0A8309BAF2E089DB3C}
         x{6D}
        x{74}
       x{B226B6CF36CF0C6_}
        x{ED44D0D401F863D20030916DE0F828D70B0A8309BAF2E089DB3C}
         x{6D}
        x{8D086002540B846967BD8AD0D9B17DFB2440228EB4C230FA9782F8A2D9B60428BEBB439C}
      x{2_}
       x{2_}
        x{2_}
         x{B24DF6CF36CF0C6_}
          x{ED44D0D401F863D20030916DE0F828D70B0A8309BAF2E089DB3C}
           x{6D}
          x{82186A2BB7D000}
         x{B245B6CF36CF0C6_}
          x{ED44D0D401F863D20030916DE0F828D70B0A8309BAF2E089DB3C}
           x{6D}
          x{8BC48656C6C6F20776F726C64218}
        x{2_}
         x{5}
          x{A975DB3CDB3C31}
           x{ED44D0D401F863D20030916DE0F828D70B0A8309BAF2E089DB3C}
            x{6D}
           x{7A}
          x{AAA2DB3CDB3C31}
           x{ED44D0D401F863D20030916DE0F828D70B0A8309BAF2E089DB3C}
            x{6D}
           x{8020}
         x{5}
          x{2_}
           x{A6A9B679B67863_}
            x{ED44D0D401F863D20030916DE0F828D70B0A8309BAF2E089DB3C}
             x{6D}
            x{75}
           x{A7A304E0B9D87ABA595CF63D09D873AC950A36CD04E13BAC6A5A56C8F96645A75E5BFBE2C47304E040AB803700CEC7299CB0755C8E4B658704E037A93BDF2A8B2DE38EA7A7BC0AB3864D04E13B2E9CB5669D96E6741D2CDB28CDD149_}
          x{AA83DB3CDB3C31}
           x{ED44D0D401F863D20030916DE0F828D70B0A8309BAF2E089DB3C}
            x{6D}
           x{8100F3}
       x{2_}
        x{2_}
         x{2_}
          x{2_}
           x{AABEED44D0D20001}
           x{AB61DB3CDB3C31}
            x{ED44D0D401F863D20030916DE0F828D70B0A8309BAF2E089DB3C}
             x{6D}
            x{8064}
          x{AF326D9E6D9E18C_}
           x{ED44D0D401F863D20030916DE0F828D70B0A8309BAF2E089DB3C}
            x{6D}
           x{8058}
         x{5}
          x{79A_}
           x{ACB6CF36CF0C6_}
            x{ED44D0D401F863D20030916DE0F828D70B0A8309BAF2E089DB3C}
             x{6D}
            x{800B}
           x{AEE3435697066733A2F2F516D625A38617576474E63656B454543537056507163476A37506868564C69697A724E384D364C515248526E6A5182_}
          x{AA45DB3CDB3C31}
           x{ED44D0D401F863D20030916DE0F828D70B0A8309BAF2E089DB3C}
            x{6D}
           x{80A6}
        x{2_}
         x{5}
          x{A9F1DB3CDB3C31}
           x{ED44D0D401F863D20030916DE0F828D70B0A8309BAF2E089DB3C}
            x{6D}
           x{6D}
          x{AA26DB3CDB3C31}
           x{ED44D0D401F863D20030916DE0F828D70B0A8309BAF2E089DB3C}
            x{6D}
           x{8058}
         x{5}
          x{A9D0DB3CDB3C31}
           x{ED44D0D401F863D20030916DE0F828D70B0A8309BAF2E089DB3C}
            x{6D}
           x{82181CA35F0E00}
          x{AA07DB3CDB3C31}
           x{ED44D0D401F863D20030916DE0F828D70B0A8309BAF2E089DB3C}
            x{6D}
           x{8D086002540B846967BD8AD0D9B17DFB2440228EB4C230FA9782F8A2D9B60428BEBB439C},
  },
}
`;
